# CREDIT LOAN RISK DATA ANALYSIS

Credits to Lao Tse for the dataset.
You can access the dataset [here](https://www.kaggle.com/datasets/laotse/credit-risk-dataset)

### Introduction

In this Jupyter notebook, the analyst used the Python programming language to analyze the following dataset about the credit risk of borrowers involved, with the intention to examine and analyze trends, findings, and what recommendations can be given based on the results. The main goal of this analysis is to examine the dataset, identify patterns and provide insights regarding the results. In this way, an analyst can then provide recommendations or suggestions that can help improve businesses.

---
### The Data Collection

The dataset consists of the said columns and their values collected from Lao Tse, a Kaggle dataset provider. The following data are presented using Python with Pandas, Matplotlib, and Numpy libraries to enable further control and examination of the values. The dataset is stored in a CSV file, and is imported in the notebook using tha Pandas library. 

The following table showcases the columns involved and their descriptions.

| Column Name | Description |
| --- | --- |
| person_age | Age of the borrower |
| person_income | Annual income |
| person_home_ownership | Status of home ownership |
| person_emp_length | Employment length in terms of years |
| loan_intent | Reason of loaning |
| loan_grade | Grade associated with the borrower or loan |
| loan_amnt | Amount of money loaned |
| loan_int_rate | Interest rate |
| loan_status | Status of the loan. ( 1 means the loan is still active) |
| loan_percent_income | Percentage of income that goes into the loan. |
| cb_person_default_on_file | If the borrower has failed to make scheduled payments|
| cb_person_cred_hist_length | Length of credit history in years |

---
### Data Cleaning and Preparation

The provided dataset is divided into separate columns, each consisting of string and numerical values. To further analyze the dataset, the analyst has gather more information using python. A summary table has also been provided.

| Column Name | Actions |
| --- | --- |
| person_age | Identified and cleaned values with invalid maximum age. Replaced with more viable values |
| person_income | Indentified and cleaned outlier value |
| person_home_ownership | No issues identified |
| person_emp_length | Null values and invalid outlier value |
| loan_intent | Spaces added between works |
| loan_grade | No issues identified |
| loan_amnt | No issues identified |
| loan_int_rate | Some values are null. Replaced using MICE |
| loan_status | Changes values from integer (1, 0) to boolean (True, False)|
| loan_percent_income | No issues identified |
| cb_person_default_on_file | Changes values from string characters (Y, N) to boolean (True, False) |
| cb_person_cred_hist_length | No issues identified |

The analyst imports the dataset and used the Pandas function of `.info()`.

In [ ]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sb

dataset = pd.read_csv('credit_risk_dataset.csv') # Import the dataset
dataset.info()

<class 'pandas.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   person_age                  32581 non-null  int64  
 1   person_income               32581 non-null  int64  
 2   person_home_ownership       32581 non-null  str    
 3   person_emp_length           31686 non-null  float64
 4   loan_intent                 32581 non-null  str    
 5   loan_grade                  32581 non-null  str    
 6   loan_amnt                   32581 non-null  int64  
 7   loan_int_rate               29465 non-null  float64
 8   loan_status                 32581 non-null  int64  
 9   loan_percent_income         32581 non-null  float64
 10  cb_person_default_on_file   32581 non-null  str    
 11  cb_person_cred_hist_length  32581 non-null  int64  
dtypes: float64(3), int64(5), str(4)
memory usage: 3.0 MB


The data above shows the general information regarding the data. It consists of 32581 entries for 12 columns. According tot he non-null count, there are some values that are null in `person_emp_length`, and `loan_int_rate`. Null means that there are no values given and does not mean that the value is 0. The data types consists of three types, int64, str, and float64. `int64` refers to whole numbers, whether negative or positive like -1, 4, 0, 99. `str` refers to values that are stored as text like RENT or MORTGAGE. `float64` refers to numbers with decimal places unlike `int64`, like 1.32, 0.00, or -13.23.

In [338]:
dataset.describe()

,person_age,person_income,person_emp_length,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_cred_hist_length
count,32581.000000,3.258100e+04,31686.000000,32581.000000,29465.000000,32581.000000,32581.000000,32581.000000
mean,27.734600,6.607485e+04,4.789686,9589.371106,11.011695,0.218164,0.170203,5.804211
std,6.348078,6.198312e+04,4.142630,6322.086646,3.240459,0.413006,0.106782,4.055001
min,20.000000,4.000000e+03,0.000000,500.000000,5.420000,0.000000,0.000000,2.000000
25%,23.000000,3.850000e+04,2.000000,5000.000000,7.900000,0.000000,0.090000,3.000000
50%,26.000000,5.500000e+04,4.000000,8000.000000,10.990000,0.000000,0.150000,4.000000
75%,30.000000,7.920000e+04,7.000000,12200.000000,13.470000,0.000000,0.230000,8.000000
max,144.000000,6.000000e+06,123.000000,35000.000000,23.220000,1.000000,0.830000,30.000000


By using `.describe()`, the dataset's summary of numerical values can be seen. This includes the number of enties, mean, standard deviation, minimum values, quartiles, and the maximum numerical values. The summary provides values that are essential in data analysis.

##### Age

This column represents the ages of the borrowers when the dataset was recorded. Analyzing this column can provide insights regarding the ages and other data. To further examine the age of the borrowers, the analyst have used the following code that targets the age of the borrowers specifically by assigning the column into a separate series of data:

In [339]:
person_ages = dataset['person_age']
person_ages.info()

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: person_age
Non-Null Count  Dtype
--------------  -----
32581 non-null  int64
dtypes: int64(1)
memory usage: 254.7 KB


The result shows the data as a series of integers consisting of 32,581 entries.

In [340]:
person_ages.describe()

count    32581.000000
mean        27.734600
std          6.348078
min         20.000000
25%         23.000000
50%         26.000000
75%         30.000000
max        144.000000
Name: person_age, dtype: float64

Using `.describe()` shows the statistical summary of the column. According to the function results, the maximum value is `144` which should be impossible for a human lifespan and treated as a data entry error. To further analyze the concern, the analyst finds the ages equivalent to `144` and the maximum age lower than the value.

In [ ]:
person_ages[person_ages != 144].max() # Get the new maximum age

np.int64(123)

Excluding `144`, another issue has occured in which the new maximum value that is not likely to occur considering human lifespan, there is also a chance that this is an entry error. That new number is `123` which is close to the oldest age a person has achieved which is 122. To analyze this further, let us examine the remaining data without the values of `123` and `144`.

In [ ]:
person_ages[(person_ages != 144) & (person_ages != 123)].max() # Get the new maximum age

np.int64(94)

Since the new possible maximum age is `94`, there is a very high chance that this is the new maximum value. This is because `94` is a valid age to get a loan like [PVAO Pension Loans](https://www.veteransbank.com.ph/pension-loan) which allows citizens 65 years old and above to qualify for a row. Now that we have a new valid maximum age, let us deal with the outliers by identifying their other details.

In [ ]:
# Get all rows that is not 144 or 123
filtered_ages = dataset[['person_age','person_income']]
filtered_ages[ (filtered_ages['person_age'] == 144) | ( filtered_ages['person_age']  == 123)]


,person_age,person_income
81,144,250000
183,144,200000
575,123,80004
747,123,78000
32297,144,6000000


In [344]:
person_ages[person_ages.eq(123)].count()

np.int64(2)

In [345]:
person_ages[person_ages.eq(144)].count()

np.int64(3)

The results indicate that the rows with `144` as the age only amount up to three rows:`81`, `183`, `32297`. On the other hand, only two rows with the value of `123` are found: `575`, `747`. Due to the low number of entry errors for age, changing their values have little effect on bias.

To treat values with `144`, the analyst have considered the following options as the new value:

> `null`: the analyst have considered to use null as the replacement value in order to not skew the statistics but having considered that all the columns are needed to be analyzed carefully, `null` will not be a suitable value in this case.

> `14`: Considering that the age is too low for a loan, and will standout as the only value lower then the minimum value found, `14` is not a possible value. The income also cannot be viable if the person is 14.

> `44`: The only possible value because it is in the range between minimum and maximum values and at the range of ages where loans are still applicable to apply.

To treat values with `123`, the analyst have considered the following options as the new value:

> `null`: the analyst have also considered to use null as the replacement value in order to not skew the statistics but having considered that all the columns are needed to be analyzed carefully, `null` will not be a suitable value in this case.

> `12 or 13`: These ages also considered too low for a loan, therefore not a suitable value. The income also cannot be viable if the person is 12 or 13.

> `21 or 31`: Also possible but very unlikely since the keyboard layout which starts from 1 then 2 and 3. If it is `21` or `31`, we are more likely to see `121` or `231`.

> `23`: The only possible value because it is also in the range between minimum and maximum values and at the range of ages where loans are still applicable to apply.

Based on the consideration and data review, it is more applicable to replace `144` with `44` and `123` with `23`. In order to change the values, the following code is used:

In [ ]:
# Replace all rows with age of 144 to 44
age_with_144 = filtered_ages['person_age'] == 144
filtered_ages.loc[age_with_144, 'person_age'] = 44
filtered_ages.iloc[[81,183,32297]]

,person_age,person_income
81,44,250000
183,44,200000
32297,44,6000000


In [ ]:
# Replace all rows with age of 123 to 23
age_with_144 = filtered_ages['person_age'] == 123
filtered_ages.loc[age_with_144, 'person_age'] = 23
filtered_ages.iloc[[575,747]]

,person_age,person_income
575,23,80004
747,23,78000


During the process, the analyst stored the rows with the age values of `123` and `144` and change the value of `'person_age'` for the said rows to match the chosen value which is `23` for `123` and `44` for `144`. Then the same rows were examined again for their new values and based on the result, the ages has changed, therefore it can be concluded that the column has been cleaned and the same changes can now be used on the main copy of the dataset with cleaned data. The cleaned column is also stored as a separate CSV file in the `cleaned_data` folder for backup. This can be done using the following code:

In [ ]:
cleaned_dataset = dataset.copy() # Make a new clean dataset
cleaned_dataset = cleaned_dataset.drop(columns=['person_age']) # Drop old column
cleaned_dataset['Clean_person_age'] = filtered_ages['person_age'] # Add a new column for cleaned values

cleaned_dataset['Clean_person_age'].to_csv('cleaned_data/cleaned_person_ages.csv', index=False) # Export

cleaned_dataset

,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,Clean_person_age
0,59000,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3,22
1,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2,21
2,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3,25
3,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2,23
4,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4,24
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,53000,MORTGAGE,1.0,PERSONAL,C,5800,13.16,0,0.11,N,30,57
32577,120000,MORTGAGE,4.0,PERSONAL,A,17625,7.49,0,0.15,N,19,54
32578,76000,RENT,3.0,HOMEIMPROVEMENT,B,35000,10.99,1,0.46,N,28,65
32579,150000,MORTGAGE,5.0,PERSONAL,B,15000,11.48,0,0.10,N,26,56


In [349]:
cleaned_dataset.iloc[[81,183,32297]]

,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,Clean_person_age
81,250000,RENT,4.0,VENTURE,C,4800,13.57,0,0.02,N,3,44
183,200000,MORTGAGE,4.0,EDUCATION,B,6000,11.86,0,0.03,N,2,44
32297,6000000,MORTGAGE,12.0,PERSONAL,C,5000,12.73,0,0.00,N,25,44


In [350]:
cleaned_dataset.iloc[[575,747]]

,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,Clean_person_age
575,80004,RENT,2.0,EDUCATION,B,20400,10.25,0,0.25,N,3,23
747,78000,RENT,7.0,VENTURE,B,20000,NaN,0,0.26,N,4,23


To wrap things up, the older column was replaced dropped and new values for their ages are not stored. The filtered column is added at the end of the table. When we take a look back to the columns with dirty values, they are now cleaned.

#### Currency

Before we head to the next column, the analyst have noticed that no currency was used. Therefore, Dollars ($) will be used on the data analysis as a stand-in for a real currency.


##### Income

This column stores the data regarding the yearly income of the borrowers when the dataset was made. Income can give many insights and is one of the most important variable when considering to provide a loan. The analyst used the following code to further examine the column's data.

In [351]:
person_incomes = dataset['person_income']
person_incomes

0         59000
1          9600
2          9600
3         65500
4         54400
          ...  
32576     53000
32577    120000
32578     76000
32579    150000
32580     42000
Name: person_income, Length: 32581, dtype: int64

In [352]:
person_incomes.info()

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: person_income
Non-Null Count  Dtype
--------------  -----
32581 non-null  int64
dtypes: int64(1)
memory usage: 254.7 KB


In [ ]:
person_incomes.describe().apply(lambda x: format(x, 'f')) 

count      32581.000000
mean       66074.848470
std        61983.119168
min         4000.000000
25%        38500.000000
50%        55000.000000
75%        79200.000000
max      6000000.000000
Name: person_income, dtype: str

The python code resulted in more detailed information regarding the column. From the data, it consists of 32581 entries and as intergers with no null values. The numerical summaries are also presented like the count, mean, standard deviation, minimum and maximum value, quartiles, and maximum values. One outlier that pops up is the maximum income which amounts to `6,000,000` and compared to the third quartile, a considerable jump in income has been observed. To further analyze the high amount of incomes, the analyst examined all the income between the third quartile and maximum income using the following code:

In [ ]:
dataset[dataset['person_income'] == 6000000] # Get all rows with income of 6000000

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
32297,144,6000000,MORTGAGE,12.0,PERSONAL,C,5000,12.73,0,0.0,N,25


In [ ]:
# Get third quartile excluding rows with income of 6000000
third_quartile_incomes = person_incomes[(person_incomes >= person_incomes.quantile(0.75)) & (person_incomes < 6000000) ]
third_quartile_incomes.describe().apply(lambda x: format(x,'f'))

count       8157.000000
mean      121808.686159
std        77703.101556
min        79200.000000
25%        89532.000000
50%       102000.000000
75%       128000.000000
max      2039784.000000
Name: person_income, dtype: str

After gathering all the values of between the third quartile and the `6,000,000` income, the new maximum income is `2039784`, which means that there is a huge gap between the largest income and the second most largest income. This is important to know as now we can treat `6,000,000` as the only outlier in its range. Therefore, it can be surmised that there is one value in the dataset in this column that can possibly an entry error. In this case, analyzing the rows of income with the same range values as the outlier can provide insights. The code below was used:

In [ ]:
# Find the statistical summary of rows that are on the same category as the row with income of 6000000
same_home_ownership=(dataset['person_home_ownership']) == 'MORTGAGE'
same_emp_length = (dataset['person_emp_length']) == 12.0
same_loan_intent = (dataset['loan_intent']) == 'PERSONAL'
same_loan_grade = (dataset['loan_grade']) == 'C'
same_loan_amount = (dataset['loan_amnt']) == '5000'
same_loan_int_rate = (dataset['loan_int_rate']) == 12.73
same_loan_status = (dataset['loan_status']) == '0'
same_loan_percent_income = (dataset['loan_percent_income']) == 0.0
same_default = (dataset['cb_person_default_on_file']) == 'N'
same_cred_hist_length = (dataset['loan_grade']) == 25

same_age_range =(dataset['person_age'] == 144) | (dataset['person_age'] == 44 )

x = dataset[(
  (same_age_range) |
  (same_home_ownership) |
  (same_emp_length) |
  (same_loan_intent) |
  (same_loan_grade) |
  (same_loan_amount) |
  (same_loan_int_rate) |
  (same_loan_status) |
  (same_loan_percent_income) |
  (same_default) |
  (same_cred_hist_length)
  ) 
  & (dataset['person_income'] != 6000000 )]
x['person_income'].describe().apply(lambda x: format(x,'f'))

count      31254.000000
mean       66342.287547
std        52867.262757
min         4000.000000
25%        39000.000000
50%        55680.000000
75%        80000.000000
max      2039784.000000
Name: person_income, dtype: str

Excluding the outlier, the income of borrowers in similar categories are relatively consistent. Therefore, the analyst concludes that the income of `6,000,000` is impossible and an entry error due to having extra zeros at the end. The analyst then considered the new value to be `60,000` removing a trailing `0`s as after analyzing, entries at similar categories are closer to that value of the mean and median, making `60,000` a viable replacement. The table below shows the entries summary:

In [357]:
dataset[(dataset['person_income'] != 6000000) & (((dataset['person_age'] >= 40) & dataset['person_age'] <= 50) & (dataset['loan_grade'] == 'C'))].describe().map('{:,.2f}'.format)

,person_age,person_income,person_emp_length,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_cred_hist_length
count,"6,457.00","6,457.00","6,320.00","6,457.00","5,827.00","6,457.00","6,457.00","6,457.00"
mean,27.78,"64,002.77",4.45,"9,214.52",13.46,0.21,0.17,5.86
std,6.40,"57,548.89",3.88,"6,044.84",0.96,0.41,0.11,4.08
min,20.00,"4,000.00",0.00,500.00,6.00,0.00,0.00,2.00
25%,23.00,"36,000.00",2.00,"4,900.00",12.98,0.00,0.09,3.00
50%,26.00,"53,000.00",4.00,"8,000.00",13.48,0.00,0.15,4.00
75%,30.00,"76,000.00",7.00,"12,000.00",13.99,0.00,0.23,8.00
max,144.00,"2,039,784.00",29.00,"35,000.00",16.11,1.00,0.71,30.00


The column can now be cleaned of dirty values and exported as its own file as backup. The old value was replaced using the following command:

In [358]:
person_incomes.iloc[32297]

np.int64(6000000)

In [ ]:
entry_error = dataset['person_income'] == 6000000 # Get all rows with income of 6000000
filtered_incomes = dataset[['person_income','person_age']] # Make a new dataset
filtered_incomes.loc[entry_error, 'person_income'] = 60000 # Replace the old value
filtered_incomes.iloc[32297] # Show row with new value


person_income    60000
person_age         144
Name: 32297, dtype: int64

In [360]:
cleaned_dataset = cleaned_dataset.drop(columns=['person_income'])
cleaned_dataset['Clean_person_income'] = filtered_incomes['person_income']
cleaned_dataset

,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,Clean_person_age,Clean_person_income
0,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3,22,59000
1,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2,21,9600
2,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3,25,9600
3,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2,23,65500
4,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4,24,54400
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,MORTGAGE,1.0,PERSONAL,C,5800,13.16,0,0.11,N,30,57,53000
32577,MORTGAGE,4.0,PERSONAL,A,17625,7.49,0,0.15,N,19,54,120000
32578,RENT,3.0,HOMEIMPROVEMENT,B,35000,10.99,1,0.46,N,28,65,76000
32579,MORTGAGE,5.0,PERSONAL,B,15000,11.48,0,0.10,N,26,56,150000


In [ ]:
cleaned_dataset['Clean_person_income'].to_csv('cleaned_data/cleaned_person_incomes.csv',index=False) # Export
person_incomes.describe().apply(lambda x: format(x, 'f'))

count      32581.000000
mean       66074.848470
std        61983.119168
min         4000.000000
25%        38500.000000
50%        55000.000000
75%        79200.000000
max      6000000.000000
Name: person_income, dtype: str

The value has been replaced and as we can clearly see, the new maximum value is no longer `6,000,000`. The cleaned column now has a backup CSV file, and on the clean dataset, old column is now dropped, replaced with the cleaned column for `person_income`.

##### Home Ownership

In this column, the collected data represents the type of home ownership borrowers have. Analyzing this column is importany as it prove to be useful in identifying trends or insights regarding their loans. To further analyze the column, the analyst used the following code:

In [362]:
home_ownerships = dataset['person_home_ownership']
home_ownerships.info()

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: person_home_ownership
Non-Null Count  Dtype
--------------  -----
32581 non-null  str  
dtypes: str(1)
memory usage: 254.7 KB


In [363]:
home_ownerships.describe()

count     32581
unique        4
top        RENT
freq      16446
Name: person_home_ownership, dtype: object

Based on the results, the columns contains string of for different types with 32581 entries. The most frequent home ownership is `RENT` with a total number of `16,446`. To identify the other kinds of ownership, the analyst used the following:

In [364]:
home_ownerships.drop_duplicates(keep='last')
# Remove all duplicates except the last copies

32291       OTHER
32572         OWN
32579    MORTGAGE
32580        RENT
Name: person_home_ownership, dtype: str

There are four types, `OTHER`, `OWN`, `MORTGAGE`, and `RENT`. Other than this, the table has no concerns or data to examine, thus the data can be considered clean. The next step follows the analyst creating a new CSV file for the column and updating the old dataset.

In [365]:
cleaned_dataset = cleaned_dataset.drop(columns=['person_home_ownership'])
cleaned_dataset['Clean_person_home_ownership'] = home_ownerships
cleaned_dataset['Clean_person_home_ownership'].to_csv('cleaned_data/cleaned_person_home_ownerships.csv', index=False)

cleaned_dataset

,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,Clean_person_age,Clean_person_income,Clean_person_home_ownership
0,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3,22,59000,RENT
1,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2,21,9600,OWN
2,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3,25,9600,MORTGAGE
3,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2,23,65500,RENT
4,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4,24,54400,RENT
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,1.0,PERSONAL,C,5800,13.16,0,0.11,N,30,57,53000,MORTGAGE
32577,4.0,PERSONAL,A,17625,7.49,0,0.15,N,19,54,120000,MORTGAGE
32578,3.0,HOMEIMPROVEMENT,B,35000,10.99,1,0.46,N,28,65,76000,RENT
32579,5.0,PERSONAL,B,15000,11.48,0,0.10,N,26,56,150000,MORTGAGE


##### Employment Length

This column represents the borrower's employment length in terms of years. The column can give insights about the borrower and information about them which can be helpful to find relationships regarding borrower information and their loan. The analyst used the following code as the next step for the analysis:

In [366]:
employment_length = dataset['person_emp_length']
employment_length.info()

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: person_emp_length
Non-Null Count  Dtype  
--------------  -----  
31686 non-null  float64
dtypes: float64(1)
memory usage: 254.7 KB


In [367]:
employment_length.describe()

count    31686.000000
mean         4.789686
std          4.142630
min          0.000000
25%          2.000000
50%          4.000000
75%          7.000000
max        123.000000
Name: person_emp_length, dtype: float64

After analyzing the column of employment length, there are two concerns: maximum value and null entries. Firstly, the maximum value is `123` years which is impossible for employment length but it also impossible to guess the proper value if we are to start from a typo error standpoint. Secondly, there are null values, only `31686` of the entries are the only one with entries including the values of `123`. Including the null and outliers, they only occupy a small percentage of the data thus filling them up with other values will not heavily affect the bias. To treat the data, the analyst focused in solving the values of improper max values. This was done using the following code:

In [368]:
employment_length[employment_length != 123.0].describe()

count    31684.000000
mean         4.782224
std          4.034888
min          0.000000
25%          2.000000
50%          4.000000
75%          7.000000
max         41.000000
Name: person_emp_length, dtype: float64

In [369]:
#Since the new max is 41.0, get all the rows with this value
dataset[employment_length == 41.0]

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
32355,78,48000,RENT,41.0,MEDICAL,A,3000,7.51,0,0.06,N,25


Excluding the outlier, the analyst have discovered that the new maximum value to be `41` and after examining further, it can be concluded that this new maximum value is the most fitting due to the other information in the column especially age of the borrower. Due to the low number of rows that has the outlier value, the new value can be replaced with the median. The same process can also be done for null values. The code below shows the process of assigning the new value:

In [370]:
median_value = employment_length[(employment_length != 123.0)].median() # Get all of median of all the values that is not 123.0
filtered_employment = dataset['person_emp_length'].replace(123.0,np.nan) # Replace all outlier with null
filtered_employment = filtered_employment.fillna(median_value) # Fill all nulls with the median value
filtered_employment.describe()


count    32581.000000
mean         4.760689
std          3.981013
min          0.000000
25%          2.000000
50%          4.000000
75%          7.000000
max         41.000000
Name: person_emp_length, dtype: float64

The nulls and outliers have been replaced with the median value and the new statistical summary shows that all values are no longer null or `123.0`. The data have been cleaned and the next process will be creating a separate CSV file and adjusting the cleaned dataset.

In [ ]:
cleaned_dataset = cleaned_dataset.drop(columns='person_emp_length') # Drop old column
cleaned_dataset['Clean_person_emp_length'] = filtered_employment # Create new column and assign new values
cleaned_dataset['Clean_person_emp_length'].to_csv('cleaned_data/cleaned_person_emp_length.csv', index=False) # Export
cleaned_dataset

,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,Clean_person_age,Clean_person_income,Clean_person_home_ownership,Clean_person_emp_length
0,PERSONAL,D,35000,16.02,1,0.59,Y,3,22,59000,RENT,4.0
1,EDUCATION,B,1000,11.14,0,0.10,N,2,21,9600,OWN,5.0
2,MEDICAL,C,5500,12.87,1,0.57,N,3,25,9600,MORTGAGE,1.0
3,MEDICAL,C,35000,15.23,1,0.53,N,2,23,65500,RENT,4.0
4,MEDICAL,C,35000,14.27,1,0.55,Y,4,24,54400,RENT,8.0
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,PERSONAL,C,5800,13.16,0,0.11,N,30,57,53000,MORTGAGE,1.0
32577,PERSONAL,A,17625,7.49,0,0.15,N,19,54,120000,MORTGAGE,4.0
32578,HOMEIMPROVEMENT,B,35000,10.99,1,0.46,N,28,65,76000,RENT,3.0
32579,PERSONAL,B,15000,11.48,0,0.10,N,26,56,150000,MORTGAGE,5.0


##### Loan Intent

This column represents the purpose of getting a loan provided by the borrower. This column is important when examining borrower information and relating it to other details in the dataset. The following column is cleaned using the process below:


In [372]:
loan_intents = dataset['loan_intent'] # load the column into a separate series
loan_intents

0               PERSONAL
1              EDUCATION
2                MEDICAL
3                MEDICAL
4                MEDICAL
              ...       
32576           PERSONAL
32577           PERSONAL
32578    HOMEIMPROVEMENT
32579           PERSONAL
32580            MEDICAL
Name: loan_intent, Length: 32581, dtype: str

In [373]:
loan_intents.info() # Get technical summary

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: loan_intent
Non-Null Count  Dtype
--------------  -----
32581 non-null  str  
dtypes: str(1)
memory usage: 254.7 KB


In [374]:
loan_intents.describe() # Get statistical summary

count         32581
unique            6
top       EDUCATION
freq           6453
Name: loan_intent, dtype: object

In [375]:
loan_intents.drop_duplicates(keep='last') #Get all unique values

32556    DEBTCONSOLIDATION
32571              VENTURE
32575            EDUCATION
32578      HOMEIMPROVEMENT
32579             PERSONAL
32580              MEDICAL
Name: loan_intent, dtype: str

Based on the output, the only concerns we have are the lack of spaces between the two words for `DEBTCONSOLIDATION` and `HOMEIMPROVEMENT`. The more desired value should be `DEBTCONSOLIDATION` and `HOME IMPROVEMENT`. To handle these strings, we can do the following:

In [376]:
to_filter_debt_intent = loan_intents == 'DEBTCONSOLIDATION' # Get all rows with 'DEBTCONSOLIDATION'
to_filter_home_improve_intent = loan_intents == 'HOMEIMPROVEMENT' # Get all rows with 'HOMEIMPROVEMENT'
loan_intents.loc[to_filter_debt_intent] = 'DEBT CONSOLIDATION' # Replace 'DEBTCONSOLIDATION' with 'DEBT CONSOLIDATION'
loan_intents.loc[to_filter_home_improve_intent] = 'HOME IMPROVEMENT' # Replace 'HOMEIMPROVEMENT' with 'HOME IMPROVEMENT'
loan_intents.drop_duplicates(keep='last') # Drop duplicates to see all unique values

32556    DEBT CONSOLIDATION
32571               VENTURE
32575             EDUCATION
32578      HOME IMPROVEMENT
32579              PERSONAL
32580               MEDICAL
Name: loan_intent, dtype: str

The values now have a space between the two words making the string look more clean. A CSV file is now made and the new column is added to the cleaned dataset.

In [377]:
cleaned_dataset = cleaned_dataset.drop(columns=['loan_intent']) #Drop the old column
cleaned_dataset['Clean_loan_intent'] = loan_intents #Add a new column fro cleaned data
cleaned_dataset['Clean_loan_intent'].to_csv('cleaned_data/cleaned_loan_intent.csv', index=False) # Create CSV file
cleaned_dataset

,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,Clean_person_age,Clean_person_income,Clean_person_home_ownership,Clean_person_emp_length,Clean_loan_intent
0,D,35000,16.02,1,0.59,Y,3,22,59000,RENT,4.0,PERSONAL
1,B,1000,11.14,0,0.10,N,2,21,9600,OWN,5.0,EDUCATION
2,C,5500,12.87,1,0.57,N,3,25,9600,MORTGAGE,1.0,MEDICAL
3,C,35000,15.23,1,0.53,N,2,23,65500,RENT,4.0,MEDICAL
4,C,35000,14.27,1,0.55,Y,4,24,54400,RENT,8.0,MEDICAL
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,C,5800,13.16,0,0.11,N,30,57,53000,MORTGAGE,1.0,PERSONAL
32577,A,17625,7.49,0,0.15,N,19,54,120000,MORTGAGE,4.0,PERSONAL
32578,B,35000,10.99,1,0.46,N,28,65,76000,RENT,3.0,HOME IMPROVEMENT
32579,B,15000,11.48,0,0.10,N,26,56,150000,MORTGAGE,5.0,PERSONAL


##### Loan Grade

This column represents the grade of the borrower, ranging from A to G. This column can be beneficial as analyzing loan grades can show insights about a borrower's loaning history and can be used to related to other information like how a grade is related to othe columns in the dataset. To examine this further, the following code was used:

In [378]:
loan_grades = dataset['loan_grade'] 
loan_grades.info()

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: loan_grade
Non-Null Count  Dtype
--------------  -----
32581 non-null  str  
dtypes: str(1)
memory usage: 254.7 KB


In [379]:
loan_grades.describe()

count     32581
unique        7
top           A
freq      10777
Name: loan_grade, dtype: object

In [380]:
loan_grades.drop_duplicates(keep='last')

32071    G
32555    F
32563    E
32574    D
32576    C
32577    A
32580    B
Name: loan_grade, dtype: str

After getting the summaries of loan_grades, it can be said that there are no issues whatsoever. Therefore, the column is already cleaned. The next process is done by using the following code:

In [381]:
# Drop the old column, add the new column to the dataset, and create the CSV file
cleaned_dataset = cleaned_dataset.drop(columns=['loan_grade'])
cleaned_dataset['Clean_loan_grade'] = loan_grades
cleaned_dataset['Clean_loan_grade'].to_csv('cleaned_data/cleaned_loan_grade.csv', index=False)

##### Loan Amount

This column represents the amount a borrower loaned from a bank or similar institution. The column is the most essential data in finding statistical insights. To clean the column, the following code was used:

In [382]:
loan_amounts = dataset['loan_amnt']
loan_amounts.info()

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: loan_amnt
Non-Null Count  Dtype
--------------  -----
32581 non-null  int64
dtypes: int64(1)
memory usage: 254.7 KB


In [383]:
loan_amounts.describe()

count    32581.000000
mean      9589.371106
std       6322.086646
min        500.000000
25%       5000.000000
50%       8000.000000
75%      12200.000000
max      35000.000000
Name: loan_amnt, dtype: float64

The output shows no issues or concerns which makes the column already cleaned and ready to be used. The next process used the following to export a CSV file and add the column in the cleaned dataset.

In [384]:
# Drop the old column, add the new column to the dataset, and create the CSV file
cleaned_dataset = cleaned_dataset.drop(columns=['loan_amnt'])
cleaned_dataset['Clean_loan_amnt'] = loan_amounts
cleaned_dataset['Clean_loan_amnt'].to_csv('cleaned_data/cleaned_loan_amnt.csv', index=False)
cleaned_dataset

,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,Clean_person_age,Clean_person_income,Clean_person_home_ownership,Clean_person_emp_length,Clean_loan_intent,Clean_loan_grade,Clean_loan_amnt
0,16.02,1,0.59,Y,3,22,59000,RENT,4.0,PERSONAL,D,35000
1,11.14,0,0.10,N,2,21,9600,OWN,5.0,EDUCATION,B,1000
2,12.87,1,0.57,N,3,25,9600,MORTGAGE,1.0,MEDICAL,C,5500
3,15.23,1,0.53,N,2,23,65500,RENT,4.0,MEDICAL,C,35000
4,14.27,1,0.55,Y,4,24,54400,RENT,8.0,MEDICAL,C,35000
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,13.16,0,0.11,N,30,57,53000,MORTGAGE,1.0,PERSONAL,C,5800
32577,7.49,0,0.15,N,19,54,120000,MORTGAGE,4.0,PERSONAL,A,17625
32578,10.99,1,0.46,N,28,65,76000,RENT,3.0,HOME IMPROVEMENT,B,35000
32579,11.48,0,0.10,N,26,56,150000,MORTGAGE,5.0,PERSONAL,B,15000


##### Loan Interest Rate

This column represents the interest rate for each loan per year. Interest rate is important for credit loaners as it is how businesses earn their profit. To further examine the data, the following code was used:

In [385]:
interest_rates = dataset['loan_int_rate']
interest_rates.info()

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: loan_int_rate
Non-Null Count  Dtype  
--------------  -----  
29465 non-null  float64
dtypes: float64(1)
memory usage: 254.7 KB


In [386]:
interest_rates.describe()

count    29465.000000
mean        11.011695
std          3.240459
min          5.420000
25%          7.900000
50%         10.990000
75%         13.470000
max         23.220000
Name: loan_int_rate, dtype: float64

The output shows that only `29465` out of the total `32581` entries have values. This column is important to be filled correctly because if the missing values are simply filled carelessly, it can skew the output. To examine the problem further, the analyst calculated the percentage of null values using the following code:

In [387]:
(interest_rates.isnull().sum() /  interest_rates.count()) * 100 # Find the percentage of null values

np.float64(10.575258781605294)

The result showed that almost `11%` of the loan interest rate data is null and randomly filling out these values will definitely skew the output. One way to solve this issue is to use MICE or Multivariate Imputation by Chained Equations. This is a technique used to handle missing data by filling it with values based on the relationship of data with other columns. To achieve this, we use the following module from `sklearn`:

In [388]:
null_rates = interest_rates[interest_rates.isnull()]
null_rates # Get the rows of null interest rates

39      NaN
50      NaN
57      NaN
59      NaN
62      NaN
         ..
32547   NaN
32552   NaN
32553   NaN
32569   NaN
32570   NaN
Name: loan_int_rate, Length: 3116, dtype: float64

In [389]:
from sklearn.experimental import enable_iterative_imputer  
from sklearn.impute import IterativeImputer

#  Dataset with missing loan values and columns to base the data from.
data = {
    'person_income': cleaned_dataset['Clean_person_income'],
    'loan_amnt': cleaned_dataset['Clean_loan_amnt'],
    'interest_rate': dataset['loan_int_rate']
}
df = pd.DataFrame(data) # Store relevant information in this dataframe

# Initialize the MICE Imputer
imputer = IterativeImputer(max_iter=12, random_state=40)
df_imputed = pd.DataFrame(imputer.fit_transform(df), columns=df.columns)
interest_rates = df_imputed['interest_rate']

In [390]:
df_imputed['interest_rate'].describe()

count    32581.000000
mean        11.011830
std          3.085959
min          3.951244
25%          8.490000
50%         10.990000
75%         13.110000
max         23.220000
Name: interest_rate, dtype: float64

After the MICE imputation, the previously nulled values were filled varying rates making the total entries of rows that has no null values to be `32581` rows. The next process involves the following code:

In [391]:
cleaned_dataset = cleaned_dataset.drop(columns=['loan_int_rate'])
cleaned_dataset['Clean_loan_int_rate'] = interest_rates
cleaned_dataset['Clean_loan_int_rate'].to_csv('cleaned_data/cleaned_loan_int_rate.csv', index=False)
cleaned_dataset

,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,Clean_person_age,Clean_person_income,Clean_person_home_ownership,Clean_person_emp_length,Clean_loan_intent,Clean_loan_grade,Clean_loan_amnt,Clean_loan_int_rate
0,1,0.59,Y,3,22,59000,RENT,4.0,PERSONAL,D,35000,16.02
1,0,0.10,N,2,21,9600,OWN,5.0,EDUCATION,B,1000,11.14
2,1,0.57,N,3,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87
3,1,0.53,N,2,23,65500,RENT,4.0,MEDICAL,C,35000,15.23
4,1,0.55,Y,4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,0,0.11,N,30,57,53000,MORTGAGE,1.0,PERSONAL,C,5800,13.16
32577,0,0.15,N,19,54,120000,MORTGAGE,4.0,PERSONAL,A,17625,7.49
32578,1,0.46,N,28,65,76000,RENT,3.0,HOME IMPROVEMENT,B,35000,10.99
32579,0,0.10,N,26,56,150000,MORTGAGE,5.0,PERSONAL,B,15000,11.48


##### Loan Status

This column represents whether the current loan is still active or not. Currently, the data is stored as `1` and `0`. The analyst transformed the following data to `True` or `False` for better analysis as boolean objects can be faster and more efficient than strings and better for human readability while still usable in machine learning. Before that, the analyst examined the data first:

In [392]:
loan_statuses = dataset['loan_status']
loan_statuses.info()

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: loan_status
Non-Null Count  Dtype
--------------  -----
32581 non-null  int64
dtypes: int64(1)
memory usage: 254.7 KB


In [393]:
loan_statuses.describe()

count    32581.000000
mean         0.218164
std          0.413006
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          1.000000
Name: loan_status, dtype: float64

The summaries shows no issues or outliers. To transform the data to `True` or `False`, the analyst used the following process:

In [394]:
loan_statuses = dataset[['loan_status']] # Get the column of loan status as datarame
true_values = loan_statuses[loan_statuses['loan_status'] == 1] # Get all rows with value of 1
false_values = loan_statuses[loan_statuses['loan_status'] == 0] # Get all rows with value of 2

 # On the new column, all rows with values of 0 has value of True
loan_statuses.loc[true_values.index, 'converted_status'] = True

 # Make a new column, all rows with values of 1 has value of False
loan_statuses.loc[false_values.index, 'converted_status'] = False


loan_statuses.head()

,loan_status,converted_status
0,1,True
1,0,False
2,1,True
3,1,True
4,1,True


The value of has now been converted to `True` or `False` and is ready to be included in the cleaned dataset and the column can now be exported as a CSV file.

In [395]:
cleaned_dataset = cleaned_dataset.drop(columns=['loan_status']) # Drop old column
cleaned_dataset['Clean_loan_status'] = loan_statuses['converted_status'] # Make new column with values of cleaned data
cleaned_dataset['Clean_loan_status'].to_csv('cleaned_data/cleaned_loan_status.csv', index=False) # Export column as a csv
cleaned_dataset

,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,Clean_person_age,Clean_person_income,Clean_person_home_ownership,Clean_person_emp_length,Clean_loan_intent,Clean_loan_grade,Clean_loan_amnt,Clean_loan_int_rate,Clean_loan_status
0,0.59,Y,3,22,59000,RENT,4.0,PERSONAL,D,35000,16.02,True
1,0.10,N,2,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,False
2,0.57,N,3,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,True
3,0.53,N,2,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,True
4,0.55,Y,4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,True
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,0.11,N,30,57,53000,MORTGAGE,1.0,PERSONAL,C,5800,13.16,False
32577,0.15,N,19,54,120000,MORTGAGE,4.0,PERSONAL,A,17625,7.49,False
32578,0.46,N,28,65,76000,RENT,3.0,HOME IMPROVEMENT,B,35000,10.99,True
32579,0.10,N,26,56,150000,MORTGAGE,5.0,PERSONAL,B,15000,11.48,False


##### Percentage of Income to Loan

This column represents the percentage of that is taken from the borrower's income in order to pay the loan. The column is important to loaners as it can be used to relate to other information in the dataset. The analyst examined the tables further using the following code:

In [396]:
income_percentages = dataset['loan_percent_income']
income_percentages.info()

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: loan_percent_income
Non-Null Count  Dtype  
--------------  -----  
32581 non-null  float64
dtypes: float64(1)
memory usage: 254.7 KB


In [397]:

income_percentages.describe()

count    32581.000000
mean         0.170203
std          0.106782
min          0.000000
25%          0.090000
50%          0.150000
75%          0.230000
max          0.830000
Name: loan_percent_income, dtype: float64

The results of the summaries show that the data are floats and are relatively clean. Since there are no more issues, the data can now be exported and added to the clean dataset:

In [398]:
cleaned_dataset = cleaned_dataset.drop(columns=['loan_percent_income'])
cleaned_dataset['Clean_loan_percent_income'] = income_percentages
cleaned_dataset['Clean_loan_percent_income'].to_csv('cleaned_data/cleaned_loan_percent_income.csv', index=False)
cleaned_dataset

,cb_person_default_on_file,cb_person_cred_hist_length,Clean_person_age,Clean_person_income,Clean_person_home_ownership,Clean_person_emp_length,Clean_loan_intent,Clean_loan_grade,Clean_loan_amnt,Clean_loan_int_rate,Clean_loan_status,Clean_loan_percent_income
0,Y,3,22,59000,RENT,4.0,PERSONAL,D,35000,16.02,True,0.59
1,N,2,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,False,0.10
2,N,3,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,True,0.57
3,N,2,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,True,0.53
4,Y,4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,True,0.55
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,N,30,57,53000,MORTGAGE,1.0,PERSONAL,C,5800,13.16,False,0.11
32577,N,19,54,120000,MORTGAGE,4.0,PERSONAL,A,17625,7.49,False,0.15
32578,N,28,65,76000,RENT,3.0,HOME IMPROVEMENT,B,35000,10.99,True,0.46
32579,N,26,56,150000,MORTGAGE,5.0,PERSONAL,B,15000,11.48,False,0.10


##### Loan Default on File

The column represents if the borrower has been defaulted on their loan. The data used is in strings of `Y` and `N`, representing if the borrower has already been defaulted. To make readability and analysis better, it is preferred to change these values into `True` or `False` boolean objects. The process involves examining the data first for other issues to proceed:

In [399]:
loan_default = dataset['cb_person_default_on_file']
loan_default.info()

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: cb_person_default_on_file
Non-Null Count  Dtype
--------------  -----
32581 non-null  str  
dtypes: str(1)
memory usage: 254.7 KB


In [400]:
loan_default.describe()

count     32581
unique        2
top           N
freq      26836
Name: cb_person_default_on_file, dtype: object

Since there are no other issues present, we can now proceed on transforming the values into `True` or `False`:

In [401]:
loan_default = dataset[['cb_person_default_on_file']]
true_values = loan_default[loan_default['cb_person_default_on_file'] == 'Y']
false_values = loan_default[loan_default['cb_person_default_on_file'] == 'N']
loan_default.loc[true_values.index, 'converted_default'] = True
loan_default.loc[false_values.index, 'converted_default'] = False
loan_default.head()

,cb_person_default_on_file,converted_default
0,Y,True
1,N,False
2,N,False
3,N,False
4,Y,True


The strings are now replaced with the boolean values. Since there are no more issues, we can now proceed to exporting the data and adding the cleaned data into dataset.

In [402]:
cleaned_dataset = cleaned_dataset.drop(columns = ['cb_person_default_on_file'])
cleaned_dataset['Clean_cb_person_default_on_file'] = loan_default['converted_default']
cleaned_dataset['Clean_cb_person_default_on_file'].to_csv('cleaned_data/cleaned_cb_person_default_on_file.csv', index = False)
cleaned_dataset

,cb_person_cred_hist_length,Clean_person_age,Clean_person_income,Clean_person_home_ownership,Clean_person_emp_length,Clean_loan_intent,Clean_loan_grade,Clean_loan_amnt,Clean_loan_int_rate,Clean_loan_status,Clean_loan_percent_income,Clean_cb_person_default_on_file
0,3,22,59000,RENT,4.0,PERSONAL,D,35000,16.02,True,0.59,True
1,2,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,False,0.10,False
2,3,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,True,0.57,False
3,2,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,True,0.53,False
4,4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,True,0.55,True
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,30,57,53000,MORTGAGE,1.0,PERSONAL,C,5800,13.16,False,0.11,False
32577,19,54,120000,MORTGAGE,4.0,PERSONAL,A,17625,7.49,False,0.15,False
32578,28,65,76000,RENT,3.0,HOME IMPROVEMENT,B,35000,10.99,True,0.46,False
32579,26,56,150000,MORTGAGE,5.0,PERSONAL,B,15000,11.48,False,0.10,False


##### Credit History Length

This columnd represents how long the loan is in terms of years. To further analyze the data, the analyst used the following code:

In [403]:
history_length = dataset['cb_person_cred_hist_length']
history_length.info()

<class 'pandas.Series'>
RangeIndex: 32581 entries, 0 to 32580
Series name: cb_person_cred_hist_length
Non-Null Count  Dtype
--------------  -----
32581 non-null  int64
dtypes: int64(1)
memory usage: 254.7 KB


In [404]:
history_length.describe()

count    32581.000000
mean         5.804211
std          4.055001
min          2.000000
25%          3.000000
50%          4.000000
75%          8.000000
max         30.000000
Name: cb_person_cred_hist_length, dtype: float64

The summaries show that there are no outliers or issues with the data. The data can now be exported as a CSV file and added to the clean dataset.

In [405]:
cleaned_dataset = cleaned_dataset.drop(columns=['cb_person_cred_hist_length'])
cleaned_dataset['Clean_cb_person_cred_hist_length'] = history_length
cleaned_dataset['Clean_cb_person_cred_hist_length'].to_csv('cleaned_data/cleaned_cb_person_cred_hist_length.csv', index=False)
cleaned_dataset


,Clean_person_age,Clean_person_income,Clean_person_home_ownership,Clean_person_emp_length,Clean_loan_intent,Clean_loan_grade,Clean_loan_amnt,Clean_loan_int_rate,Clean_loan_status,Clean_loan_percent_income,Clean_cb_person_default_on_file,Clean_cb_person_cred_hist_length
0,22,59000,RENT,4.0,PERSONAL,D,35000,16.02,True,0.59,True,3
1,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,False,0.10,False,2
2,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,True,0.57,False,3
3,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,True,0.53,False,2
4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,True,0.55,True,4
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,57,53000,MORTGAGE,1.0,PERSONAL,C,5800,13.16,False,0.11,False,30
32577,54,120000,MORTGAGE,4.0,PERSONAL,A,17625,7.49,False,0.15,False,19
32578,65,76000,RENT,3.0,HOME IMPROVEMENT,B,35000,10.99,True,0.46,False,28
32579,56,150000,MORTGAGE,5.0,PERSONAL,B,15000,11.48,False,0.10,False,26


After cleaning all the data, the data set can now be considered prepared for Data Exploration and Data Analysis. For the last step, the analyst exported the cleaned dataset as a separate CSV file.

In [406]:
cleaned_dataset.to_csv('cleaned_credit_risk_dataset.csv',index=False)